From: Yinghai Lu <yinghai@kernel.org>
Subject: x86: fix sci on ioapic 1
Patch-Mainline: not yet, should still go into 2.6.33
References: bnc#572661

http://lkml.org/lkml/2010/2/2/87

Thomas Renninger <trenn@suse.de> reported on IBM x3330

booting a latest kernel on this machine results in:

PCI: PCI BIOS revision 2.10 entry at 0xfd61c, last bus=1
PCI: Using configuration type 1 for base access bio: create slab <bio-0> at 0
ACPI: SCI (IRQ30) allocation failed
ACPI Exception: AE_NOT_ACQUIRED, Unable to install System Control Interrupt handler (20090903/evevent-161)
ACPI: Unable to start the ACPI Interpreter

Later all kind of devices fail...

and bisect it down to this commit:
commit b9c61b70075c87a8612624736faf4a2de5b1ed30

    x86/pci: update pirq_enable_irq() to setup io apic routing

it turns out we need to set irq routing for the sci on ioapic1 early.

-v2: make it work without sparseirq too.

Reported-by: Thomas Renninger <trenn@suse.de>
Signed-off-by: Yinghai Lu <yinghai@kernel.org>
Signed-off-by: Thomas Renninger <trenn@suse.de>

Automatically created from "patches.fixes/x86_irq_setup_extra_ioapic_for_sci.patch" by xen-port-patches.py

--- sle11sp1-2010-02-02.orig/arch/x86/kernel/acpi/boot.c	2009-11-06 10:52:02.000000000 +0100
+++ sle11sp1-2010-02-02/arch/x86/kernel/acpi/boot.c	2010-02-02 15:08:30.000000000 +0100
@@ -455,7 +455,7 @@ int acpi_gsi_to_irq(u32 gsi, unsigned in
 {
 	*irq = gsi;
 
-#ifdef CONFIG_X86_IO_APIC
+#if defined(CONFIG_X86_IO_APIC) && !defined(CONFIG_XEN)
 	if (acpi_irq_model == ACPI_IRQ_MODEL_IOAPIC)
 		setup_IO_APIC_irq_extra(gsi);
 #endif
